core/src/main/scala/org/apache/spark/util/ManualClock.scala

```diff
@@ -27,6 +27,7 @@ package org.apache.spark.util
 private[spark] class ManualClock(private var time: Long) extends Clock {
 
   private var _isWaiting = false
+  private var _readyForFirstPeek = false
```

Comment:
I think semantically this ManualClock is becoming a very complex and confusing API to use. The caller has to 

How about changing the API to something like this. 

- ManualClock: It has a method called `isThreadWaitingAt(timeWhenWaitStarted: Long): Boolean`. It returns true when another thread has started waiting when the time was `timeWhenWaitStarted`. This replaces `isWaiting`.

- StreamTest: It keeps track of the expected time when manual clock is being used. For every AdvaneManualClock is first calls `isThreadWaitingAt(expectedTime)` until it returns true, then advances manual clock as well as the expected time.

This will ensure that successive AdvanceManualClock will wait for the StreamExecution thread to be unblocked from the previous wait, and start a new wait on the expected time.

How does this sound?

If this sounds good, then there should be unit tests to test this Manual Clock behavior thoroughly.

-